### PR TITLE
Rework contribution score function backend

### PR DESCRIPTION
**Pull request by casblaauw**
## Background
When trying to explain even one sequence from Borzoi, the explaining functions would always OOM. Digging a bit revealed `integrated_grad` makes a batch of 26 sequences to predict it and the background, which is never gonna work for a massive model like this. There was a preliminary batching function in the explainer code, but it wasn't accessible from the CREsted wrapper, and (at least on TensorFlow) its implementation gave me a lot of trouble. While fixing this, a combination of chasing speedups and frustration with code duplication led to a big rework of the explainer functions. 

### Structure
I cleaned up the function structure, separating only a few 'primitive' functions that operate on one sequence/batch to directly calculate gradients (and the batching function `function_batch()`) as backend-specific functions. The rest of the functions, which do more high-level creating backgrounds/averaging over gradients/etc are shared between tensorflow and pytorch.
To do this batching, I now reworked the batching function, added it to all functions, and exposed `batch_size` from CREsted's `tl.contribution_scores()`. 

### Integrated gradients
Default code should work exactly the same for 'integrated_grad': there, it's always processed every sequence individually and turned it into size 26 batch inside every `integrated_grad()` call, which is far under the default 128 batch size. If it's under the batch size, the function is called without any batching. I considered integrating this over sequences so that we can batch multiple sequences in one run, but that led to some bugs, and would possibly lead to problems on the expected integrated gradient side (saving 26*500 gradients might still work, but saving 26*25*500=325000 could very well cause memory issues, and then we'd need CPU-level batching as well, and blegh. Integrated gradients are plenty fast already.)

### Expected integrated gradients
For 'expected_integrated_grad', it previously called `integrated_grad()` 25 times (once for each baseline). I've simplified this and integrated it in the default `integrated_grad()` function. Therefore, instead of 25 calls of 26 sequences ( `n_steps+1` sequences for `n_baselines` baselines) it now creates 650 sequences (`n_baselines`*(`n_steps+1`)) in one go and only needs 6 batches of 128 seqs.
Also, previously, the shuffling didn't accept a seed value. I've changed that so that expected integrated gradients should be more consistent now.

## Other mini changes
- I exposed saliency maps (aka raw gradients) from `tl.contribution_scores()` - the recommended technique should still be `expected_integrated_grad` but I can see people need this since it's way way faster especially when you're working with low batch sizes (since expected_integrated_grad takes 25*26=650 sequence explanations to explain one sequence, which hurts when that's 650 gradient tracking predictions instead of 25 batches of 26). More selfishly, I needed it for the Borzoi comparisons, and it's implemented anyway so might as well expose it. 
- Also, I added mutagenesis to the test suite - if it's too slow we can remove it again, but it is an important function that wasn't being tested.

## Review
- I haven't tested this on PyTorch - I trust the CI will catch anything wrong, but an extra look over the code on that side is especially welcome

## Further changes
- A massive speedup would be to implement Seppe's idea of getting gradients for multiple classes in the `saliency_map` function, rather than running everything again per class - I don't have time to implement that atm. 
- Mutagenesis code probably deserves a look, I didn't touch that beyond making it accept batch_size atm.

### CODE REVIEW

**Comment by casblaauw** [comment]
Note: this is related to #104 , just tagging it here. This could be a way to fix the OOM issues there, or get both speedups and keep memory load managable?

Also, I have zero clue why it's failing on a loss function not being serialised, I didn't touch any of that. Problem for tomorrow.

**Comment by casblaauw** [comment]
Some speed and equivalence tests:

Raw saliency maps:
(Compares with prev also-batched implementation, which gave a warning every time it's used)
```
DilatedCNN model (500 seqs): old 0.69s vs new 0.69s
DilatedCNN model (500 seqs) old vs new saliencymap equal: True, allclose True
```

Integrated gradients:
```
DilatedCNN model (3 seqs): old 0.05s vs new 0.05s
DilatedCNN model (3 seqs) old new integrated grads: equal: False, allclose atol 1e-6 True
DilatedCNN model (500 seqs): old 26.22s vs new 25.31s
Borzoi model (3 seqs): old integrated gradients impossible vs new integrated grads 29.04s
```

Expected integrated gradients
```
DilatedCNN model (3 seqs): old exp int grads 3.91 vs new exp int grads 2.82
DilatedCNN model (100 seqs): old exp int grads 130.75 vs new exp int grads 94.12
DilatedCNN model (1 seq) old exp int grads vs new exp int grads (batch size 128) equal: False, allclose atol 1e-05 True
```

Some of the sources of variance:
- In expected integrate gradients, sequence inputs were exactly the same, but running them in 2 batches of 26 vs 1 batch of 52 gave ever so slightly different gradients. 1e-7 level stuff, so not detectable in practice.
- Similarly (but even more minor), saving arrays to a list and concatenating vs writing them to a slice of a pre-made array led to miniscule differences.
For each of these 3, I checked the results from explaining one sequence, and they look exactly the same when plotted. The previous variance of setting a seed for expected integrated gradients will have been much higher.

**Comment by casblaauw** [comment]
My tests are failing on an unrelated issue to the PR: `CosineMSELogLoss` not being decorated with `@keras.saving.register_keras_serializable(package="Losses")`, which means a model can't be loaded in with compiled=True if it used that loss. While it's true that that function doesn't have that decorator, unlike `CosineMSELoss`, this has always been the case... The test that's failing is also at least 3 months old.

I've now added it as a commit, no clue why it decided to complain now and not in the past few months but okay.

**Comment by casblaauw** [comment]
Tagging @LukasMahieu and @SeppeDeWinter for a look over this, if at all possible also testing the function on pytorch if you have an environment ready.

File: src/crested/tl/_explainer_tf.py [hunk 1]
```diff
@@ -139,119 +53,30 @@ def smoothgrad(
...
-    x_shuffle = []
-    for _ in range(num_samples):
-        shuffle = np.random.permutation(x.shape[1])
-        x_shuffle.append(x[0, shuffle, :])
-    return np.array(x_shuffle)
+    data_size = X.shape[0]
+    if data_size <= batch_size:
+        return fun(X, **kwargs).numpy()
+    else:
+        outputs = np.zeros_like(X)
```
Comment (SeppeDeWinter) [inline src/crested/tl/_explainer_tf.py:74]: This could also be a tensor instead, that way you don't have to call `.numpy()` for each batch.
You can call it once when returning `return outputs.numpy()`. Could be a slight performance boost. The overhead of `.numpy()` is probably not that large, I would guess?

File: src/crested/tl/_explainer_torch.py [hunk 2]
```diff
@@ -124,132 +36,30 @@ def smoothgrad(
...
-def random_shuffle(x, num_samples=1):
-    """Randomly shuffle sequences. Assumes x shape is (N,L,A)."""
-    x_shuffle = []
-    for _ in range(num_samples):
-        shuffle = np.random.permutation(x.shape[1])
-        x_shuffle.append(x[0, shuffle, :])
-    return np.array(x_shuffle, dtype=np.float32)
+    data_size = X.shape[0]
+    if data_size <= batch_size:
+        return fun(X, **kwargs).detach().cpu().numpy()
```
Comment (SeppeDeWinter) [inline src/crested/tl/_explainer_torch.py:55]: Same comment here as in `_explainer_tf.py`. This assumes that the function always returns a `torch.tensor`. I would again make this explicit in either the docstring or by annotation the type of `fun`.

File: src/crested/tl/_explainer_torch.py [hunk 3]
```diff
@@ -124,132 +36,30 @@ def smoothgrad(
...
+    else:
+        outputs = np.zeros_like(X)
```
Comment (SeppeDeWinter) [inline src/crested/tl/_explainer_torch.py:57]: Same comment as in `_explainer_tf.py`. Data can be copied to CPU once when returning, instead of for each batch.
Or, are we worried about filling up GPU memory? Then your approach is probably indeed best.

File: src/crested/tl/_tools.py [hunk 4]
```diff
@@ -342,22 +346,39 @@ def contribution_scores(
...
-                scores[:, i, :, :] = explainer.mutagenesis(
+                scores[:, i, :, :] = mutagenesis(
                     input_sequences,
+                    model=m,
                     class_index=class_index,
+                    batch_size=batch_size
                 )
             elif method == "expected_integrated_grad":
-                scores[:, i, :, :] = explainer.expected_integrated_grad(
+                scores[:, i, :, :] = integrated_grad(
```
Comment (SeppeDeWinter) [inline src/crested/tl/_tools.py:367]: I think you missed the `model` argument here.

**Review by SeppeDeWinter — COMMENTED** [review]
Code overal looks good to me.
Some minor comments.

Also I'm a big fan of type annotation, but given that these are missing from many more functions that would warrant another PR.

File: src/crested/tl/_explainer_torch.py (hunk 2 above)
Comment (SeppeDeWinter) [reply src/crested/tl/_explainer_torch.py:55]: Another idea is to make `_function_batch` generic to both tensorflow and pytorch.

**Comment by casblaauw** [comment]
I've added type annotation and more docs, you're right - it's a small hassle, and at least adding them every time we go over a function will increase coverage over time.

For the transfers to/from the GPU, I did originally write this to minimise GPU memory, but handling the entire matrix in GPU (i.e. `(pred(batch) for batch in batches).numpy()` vs `(pred(batches).numpy() for batch in batches)`) would probably be a bit of a speedup. I do still like transferring to CPU/numpy somewhat quickly, simply because almost any tensor-handling code will need to be duplicated for tensorflow and pytorch, which is horrible to maintain.

Okay, I tried a quick implementation, but a few tests show that it doesn't really change anything (I think):
```
# Original CREsted implementation 
Integrated grads, 500 seqs: old old 26.22s
Expected integrated grads, 100 seqs: old old 131.55s

# Current branch: move each batch separately and insert in array
Integrated grads, 500 seqs: move each batch, write to numpy array: 25.30s
Expected integrated grads, 100 seqs: move each batch, write to numpy array: 91.22

# New: save to list, concat with tf.concat, then move to numpy:
Integrated grads, 500 seqs: concatenating list before transferring: 25.26s
Expected integrated grads, 100 seqs: concatenating list before transferring: 91.39

# New: save to tf.Variable array, then move to numpy:
Integrated grads, 500 seqs: writing to tf.variable array: 25.21s
Expected integrated grads, 100 seqs:  writing to tf.variable array: 90.97
```

So I think I'll keep it as is, since that 1. reduces GPU memory load and 2. provides a clear 'separation' between shared numpy code and the backend-specific functions. It might lead to torch speedups though, and maybe doing the entire expected integrated grad integration/averaging on GPU before transferring would speed up more (transferring `(1, seq_len, 4)` instead of `(num_baselines, num_steps, seq_len, 4)`). Question for a later PR.

**Comment by casblaauw** [comment]
Tested both `crested.tl.contribution_scores` and underlying functions in PyTorch with the new code, works and gives equivalent results to with TensorFlow (when judged by plotting scatterplots of scores on top of eachother). Merging this.
